File: KODE/ADA_KLASSER/BSTOLE/BinarySearchTree.cpp
#include "BinarySearchTree.h"


BinarySearchTree::BinarySearchTree() : root{ nullptr }
{
}


BinarySearchTree::BinarySearchTree(const BinarySearchTree & rhs) : root{ nullptr }
{
	root = clone(rhs.root);
}


BinarySearchTree::BinarySearchTree(BinarySearchTree && rhs) : root{ rhs.root }
{
	rhs.root = nullptr;
}

BinarySearchTree::~BinarySearchTree()
{
	makeEmpty();
}

const int & BinarySearchTree::findMin() const
{
	if (isEmpty())
		return -1;
	return findMin(root)->element;
}


const int & BinarySearchTree::findMax() const
{
	if (isEmpty())
		return -1;
	return findMax(root)->element;
}

bool BinarySearchTree::contains(const int & x) const
{
	return contains(x, root);
}

bool BinarySearchTree::isEmpty() const
{
	return root == nullptr;
}

/*void BinarySearchTree::printTree() const
{
	if (isEmpty())
		cout << "Empty tree" << endl;
	else
		printTree(root);
}*/

void BinarySearchTree::makeEmpty()
{
	makeEmpty(root);
}


void BinarySearchTree::insert(const int & x)
{
	insert(x, root);
}

void BinarySearchTree::insert(int && x)
{
	insert(std::move(x), root);
}

void BinarySearchTree::remove(const int & x)
{
	remove(x, root);
}

void BinarySearchTree::insert(const int & x, BinaryNode * & t)
{
	if (t == nullptr)
		t = new BinaryNode{ x, nullptr, nullptr };
	else if (x < t->element)
		insert(x, t->left);
	else if (t->element < x)
		insert(x, t->right);
	else
		;  // Duplicate; do nothing
}


void BinarySearchTree::insert(int && x, BinaryNode * & t)
{
	if (t == nullptr)
		t = new BinaryNode{ std::move(x), nullptr, nullptr };
	else if (x < t->element)
		insert(std::move(x), t->left);
	else if (t->element < x)
		insert(std::move(x), t->right);
	else
		;  // Duplicate; do nothing
}

void BinarySearchTree::remove(const int & x, BinaryNode * & t)
{
	if (t == nullptr)
		return;   // Item not found; do nothing
	if (x < t->element)
		remove(x, t->left);
	else if (t->element < x)
		remove(x, t->right);
	else if (t->left != nullptr && t->right != nullptr) // Two children
	{
		t->element = findMin(t->right)->element;
		remove(t->element, t->right);
	}
	else
	{
		BinaryNode *oldNode = t;
		t = (t->left != nullptr) ? t->left : t->right;
		delete oldNode;
	}
}

BinaryNode * BinarySearchTree::findMin(BinaryNode *t) const
{
	if (t == nullptr)
		return nullptr;
	if (t->left == nullptr)
		return t;
	return findMin(t->left);
}

BinaryNode * BinarySearchTree::findMax(BinaryNode *t) const
{
	if (t != nullptr)
		while (t->right != nullptr)
			t = t->right;
	return t;
}

bool BinarySearchTree::contains(const int & x, BinaryNode *t) const
{
	if (t == nullptr)
		return false;
	else if (x < t->element)
		return contains(x, t->left);
	else if (t->element < x)
		return contains(x, t->right);
	else
		return true;    // Match
}

void BinarySearchTree::makeEmpty(BinaryNode * & t)
{
	if (t != nullptr)
	{
		makeEmpty(t->left);
		makeEmpty(t->right);
		delete t;
	}
	t = nullptr;
}


/*void BinarySearchTree::printTree(BinaryNode *t) const
{
	if (t != nullptr)
	{
		printTree(t->left);
		cout << t->element << endl;
		printTree(t->right);
	}
}
*/
void BinarySearchTree::printTree() const
{
    if (!root) {
        std::cout << "(empty)\n";
        return;
    }
    printTreeRec(root, 0);
}

BinaryNode * BinarySearchTree::clone(BinaryNode *t) const
{
	if (t == nullptr)
		return nullptr;
	else
		return new BinaryNode{ t->element, clone(t->left), clone(t->right) };
}

//-------------access points to start from root----------------------
void BinarySearchTree::inorderTraversal() const //const so we can call it like tree.inorderTraversal from main
{
    inorderTraversal(root);
}

void BinarySearchTree::preorderTraversal() const
{
    preorderTraversal(root);
}

void BinarySearchTree::postorderTraversal() const
{
    postorderTraversal(root);
}

//------------------recursive private member functions---------------------

// Function to do inorder traversal of BST
void BinarySearchTree::inorderTraversal(BinaryNode *t) const //const so we can call from public accessors
{
    if (t != nullptr) {
        inorderTraversal(t->left);
        std::cout << t->element << " ";
        inorderTraversal(t->right);
    }
}

// Function to do preorder traversal of BST
void BinarySearchTree::preorderTraversal(BinaryNode *t) const
{   
    if(!t)return;
    std::cout <<t->element << " ";
    preorderTraversal(t->left);        
    preorderTraversal(t->right);
            

}

// Function to do postorder traversal of BST
void BinarySearchTree::postorderTraversal(BinaryNode *t) const
{   
    if(!t)return;
    postorderTraversal(t->left);        
    postorderTraversal(t->right);
    std::cout <<t->element << " ";            
    
}

string BinarySearchTree::findRoute(int value) const
{ 
    if (!contains(value) || root == nullptr) {
        return "";            
    }

  
    return findRoute(root, value);  // use and return what helper gives
}

string BinarySearchTree::findRoute(BinaryNode* node, int value) const{
    
	if (node == nullptr){
		return ""; //base case
	} 
	if (value == node->element) {
		return std::to_string(node->element);
	}

	std::string route;

	if (value < node->element){
		route = std::to_string(node->element) + " " + findRoute(node->left, value);
	} else {
		route = std::to_string(node->element) + " " + findRoute(node->right, value);
	} 

	return route;
} 


int BinarySearchTree::getTreeHeight() const {
    return getTreeHeight(root);
}

int BinarySearchTree::getTreeHeight(BinaryNode* node) const {
    if (node == nullptr) {
        return -1;  // empty node, go 1 step back 
    }

    int leftHeight  = getTreeHeight(node->left);
    int rightHeight = getTreeHeight(node->right);

    return 1 + (leftHeight > rightHeight ? leftHeight : rightHeight); //root is 1
}

int BinarySearchTree::getInternalDepth() const {
    return getInternalDepth(root, 0);   // root has depth 0
}

int BinarySearchTree::getInternalDepth(BinaryNode* node, int depth) const {
    if (node == nullptr) {
        return 0;
    }

	depth = depth + getInternalDepth(node->left,  depth + 1) + getInternalDepth(node->right, depth + 1);	

    return depth;
}

int BinarySearchTree::height() const {
	return height(root);
}

int BinarySearchTree::height(BinaryNode *t) const{
	if (t == nullptr)
		return -1;
	else
		return 1 + max(height(t->left), height(t->right));
}

int BinarySearchTree::noOfNodes()const {
	return noOfNodes(root);
}

int BinarySearchTree::noOfNodes(BinaryNode *t) const{
	if (t == nullptr)
		return 0;
	else
		return 1 + noOfNodes(t->left) + noOfNodes(t->right);
}

bool BinarySearchTree::minimalHeight() const{
	   return minimalHeight(root);
}

bool BinarySearchTree::minimalHeight(BinaryNode *t) const{
	//find tree height
	int h = height();
	
	//find number of nodes
	int n = noOfNodes();
	
	std::cout<<"n: "<<n<<std::endl;
	int p = 1; //2-tals-potens
	int k = 0; //eksponent [tæller]

	//find next power of two
   	while (p < n+1) { //potens skal bare være mindre end number of nodes
    p *= 2;           // næste 2-potens
    ++k;              // eksponent stiger
	}
	std::cout<<"p: "<<p<<std::endl;
	
	//subtract one
	k = k-1;
	std::cout<<"k: "<<k<<std::endl;

	//compare to height
	if(h == k){
		return true;
	}
	return false;
}

File: ADA_MATERIALER/Kode/BinarySearchTree.java
import java.lang.Math;
/**
 * Implements an unbalanced binary search tree.
 * inspired by Mark Allen Weiss
 */
public class BinarySearchTree
{
    public BinarySearchTree( )
    {
        root = null;
    }

    // Insert into the tree; duplicates are ignored.
    public void insert( int x )
    {
        root = insert( x, root );
    }
    
    public void remove( int x )
    {
        root = remove( x, root );
    }
    
    public BinaryNode findMin()
    {
        if( isEmpty( ) )
            ;
        return findMin( root );
    }
    
    public BinaryNode findMax()
    {
        if( isEmpty( ) )
            ;
        return findMax( root);
    }

    public boolean contains( int x )
    {
        return contains( x, root );
    }
    
    public void makeEmpty( )
    {
        root = null;
    }
    
    public boolean isEmpty( )
    {
        return root == null;
    }

    public void printTree( )
    {
        if( isEmpty( ) )
            System.out.println( "Empty tree" );
        else
            printTree( root );
    }
    
    private BinaryNode insert( int x, BinaryNode t )
    {
        if( t == null )
            return new BinaryNode( x, null, null );
        
        if( x < t.element )
            t.left = insert( x, t.left );
        else
        if( x > t.element )
            t.right = insert( x, t.right );
        else
            ;  // Duplicate; do nothing
        return t;
    }
    
    private BinaryNode remove( int x, BinaryNode t )
    {
        if( t == null )
            return t;   // Item not found; do nothing
            
        if( x < t.element )
            t.left = remove( x, t.left );
        else
        if( x > t.element )
            t.right = remove( x, t.right );
        else
        if( t.left != null && t.right != null ) // Two children
        {
            t.element = findMin( t.right ).element;
            t.right = remove( t.element, t.right );
        }
        else
            t = ( t.left != null ) ? t.left : t.right;
        return t;
    }
    
    private BinaryNode findMin( BinaryNode t)
    {
        if( t == null )
            return null;
        else if( t.left == null )
            return t;
        return findMin( t.left );
    }
    
    private BinaryNode findMax( BinaryNode t )
    {
        if( t != null )
            while( t.right != null )
                t = t.right;

        return t;
    }
    
    private boolean contains( int x, BinaryNode t )
    {
        if( t == null )
            return false;
        if (x < t.element )
            return contains( x, t.left );
        else
        if( x > t.element )
            return contains( x, t.right );
        else
            return true;    // Match
    }
    
    private void printTree( BinaryNode t)
    {
        if( t != null )
        {
            printTree( t.left );
            System.out.println( t.element );
            printTree( t.right );
        }
    }

    private int height( BinaryNode t )
    {
        if( t == null )
            return -1;
        else
            return 1 + Math.max( height( t.left ), height( t.right ) );    
    }

    // Basic node stored in unbalanced binary search trees
    private static class BinaryNode
    {
            // Constructors
        BinaryNode( int theElement )
        {
            this( theElement, null, null );
        }

        BinaryNode( int theElement, BinaryNode lt, BinaryNode rt )
        {
            element  = theElement;
            left     = lt;
            right    = rt;
        }

        int element;            // The data in the node
        BinaryNode left;   // Left child
        BinaryNode right;  // Right child
    }


      /** The tree root. */
    private BinaryNode root;


        // Test program
    public static void main( String [ ] args ) {
        BinarySearchTree t = new BinarySearchTree();


        t.insert(7);
        t.insert(4);
        t.insert(28);
        t.insert(3);
        t.insert(55);
        t.insert(2);
        t.insert(51);
        t.insert(60);
        t.insert(1);
        t.insert(48);
        t.insert(58);
        t.insert(69);
        t.insert(40);
        t.insert(57);
        t.insert(35);
        t.remove(55);
        t.remove(3);
        System.out.println(t.contains(80));
        t.printTree();
    }
}
